File: 85/solution.cpp
// monotonic stack

class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        vector<int> heights;
        int result = 0;
        
        for(int i = 0; i < matrix.size(); i++) {
            heights.resize(matrix[i].size());
            for(int j = 0; j < matrix[i].size(); j++) {
                heights[j] = matrix[i][j] == '0' ? 0 : (1 + heights[j]);
            }
            result = max(result, largestRectangleArea(heights));
        }
        
        return result;
    }
    
private:
    int largestRectangleArea(vector<int> &heights) {
        heights.push_back(0);
        int n = heights.size();
        stack<int> s;
        int result = 0, i = 0;
        
        while(i < n) {
            if(s.empty() || heights[i] > heights[s.top()]) {
                s.push(i);
                i++;
            } else {
                int h = heights[s.top()]; s.pop();
                int w = s.empty() ? i : i - s.top() - 1;
                result = max(result, h * w);
            }
        }
        
        return result;
    }
};

File: 85/solution.py
# monotonic stack

class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return 0
        heights = [0] * len(matrix[0])
        result = 0
        
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                heights[j] = 0 if matrix[i][j] == '0' else (1 + heights[j])
            result = max(result, self.largestRectangleArea(heights))
        
        return result
    
    def largestRectangleArea(self, h: List[int]):
        n = len(h) + 1
        heights = [0] * n
        for i in range(n - 1):
            heights[i] = h[i]
        heights[-1] = 0
        s = []
        result, i = 0, 0
        
        while i < n:
            if not s or heights[i] >= heights[s[-1]]:
                s.append(i)
                i += 1
            else:
                h = heights[s[-1]]; s.pop()
                w = i if not s else i - s[-1] - 1
                result = max(result, h * w)
                
        return result
